packages/web3-providers-ws/src/index.js

```diff
@@ -250,36 +247,42 @@ WebsocketProvider.prototype._addResponseCallback = function(payload, callback) {
  @method _timeout
  */
 WebsocketProvider.prototype._timeout = function() {
-    for(var key in this.responseCallbacks) {
-        if(this.responseCallbacks.hasOwnProperty(key)){
+    for (var key in this.responseCallbacks) {
+        if (this.responseCallbacks.hasOwnProperty(key)) {
             this.responseCallbacks[key](errors.InvalidConnection('on WS'));
             delete this.responseCallbacks[key];
         }
     }
 };
 
 
-WebsocketProvider.prototype.send = function (payload, callback) {
+WebsocketProvider.prototype.send = function(payload, callback) {
     var _this = this;
 
     if (this.connection.readyState === this.connection.CONNECTING) {
-        setTimeout(function () {
-            _this.send(payload, callback);
-        }, 10);
+        this.connection.addEventListener(
+            'open',
```

Comment:
This is true it will with the changed lines of code just hang until the connection got established and in the case of a non-existing host will it never execute the JSON-RPC methods we have in the pipeline. The emitted error ``net::ERR_CONNECTION_REFUSED`` which would signal us the non-existing host. Can't get caught with a try/catch in the ``WebsocketProvider`` constructor because it doesn't throw an error in the normal sense. The ``WebSocket`` class does dispatch a non-listenable event because of the way it is implemented. A solution would be to create a ``connecting-timeout`` function in the constructor of the ``WebsocketProvider`` which throws an error if the connection doesn't get established in the defined time frame.

Reply:
> A solution would be to create a connecting-timeout function in the constructor of the WebsocketProvider which throws an error if the connection doesn't get established in the defined time frame.

Great analysis! To me it seems like there are two execution contexts for Web3 with opposite expectations:
+ NodeJS (testing/tools) --> fail fast and hard.
+ Network (browser/dapps) --> keep trying for as long as it takes
 
At the moment, web3 behaves mostly the way NodeJS contexts would prefer, and there are lots of complaints on the browser / dapp side. Maybe we can just acknowledge the difference in the docs/release notes and address it by toggling the different behaviors  via  the `autoReconnect: true`  (or something similar) proposed in #3167. Here you might just check that flag - stay open if true, fail fast as Web3 currently does if false?

WDYT? This change could be paired with #3167...